### DISCUSSION TITLE
Matter PWM devices show in Google Home (GH)  but act as one in GH

### DISCUSSION THREAD

**Discussion by ivanvachovsky** [arendst/Tasmota#22289] [Q&A]
Running Tasmota 14.2.0 on ESP32 board. I have configured 8 PWM devices and activated them in Google Home through Tasmota/Matter integration. They do show in GH as dimmers, which is correct. However when I set one device to say PWM3 to 30% than all 8 GH respective devices become 30%. Furthermore, the PWM=30% do not show in Tasmota. The good news is that the MQTT report is correct in two ways. Only PWM3 is reported (not all) as set to 30% and the dim level 30% is correct as well as the POWER=1. Here is the log

05:02:17.197 MQT: stat/VoiceBridge1/COMMAND = {"MtrReceived":{"3":{"Ep":3,"Bri":76,"Dimmer":30,"Power":1}}}
05:02:40.616 CMD: PWM3
05:02:40.621 MQT: stat/VoiceBridge1/RESULT = {"PWM":{"PWM1":0,"PWM2":0,"PWM3":0,"PWM4":39,"PWM5":0,"PWM6":0,"PWM7":0,"PWM8":0,"PWM9":0}}

**Comment by s-hadinger** [comment]
Please set log level to 3 with `weblog 3` or `seriallog 3` and report logs of all scenarios : changing locally and changing from GH

**Comment by ivanvachovsky** [comment]
Here it is

```
06:55:48.202 CMD: weblog 3
06:55:48.207 MQT: stat/VoiceBridge1/RESULT = {"WebLog":3}
06:55:49.410 CFG: Saved, Count 124, Bytes 4096
06:56:00.826 MTR: <Sub_Alive (   887) sub=2041
06:56:10.702 MQT: stat/VoiceBridge1/COMMAND = {"MtrReceived":{"8":{"Ep":8,"Bri":160,"Dimmer":63,"Power":1}}}
06:56:10.709 MTR: >Command   (   887) [08]0008/0004 MoveToLevelWithOnOff (bri:160) 
06:56:10.712 MTR: <Replied   (   887) OK exch=30269
06:56:10.740 MTR: <Sub_Data  (   887) sub=2041
06:56:10.773 MTR: >Read_Attr (   887) [08]0008/0000 (CurrentLevel) - 160U
06:56:10.854 MTR: .Sub_Done  (      ) sub=2041
06:56:10.871 MTR: <Sub_Data  (   887) sub=2041
06:56:10.891 MTR: >Read_Attr (   887) [02]0008/0000 (CurrentLevel) - 160U
06:56:10.904 MTR: >Read_Attr (   887) [03]0008/0000 (CurrentLevel) - 160U
06:56:10.917 MTR: >Read_Attr (   887) [04]0008/0000 (CurrentLevel) - 160U
06:56:10.931 MTR: >Read_Attr (   887) [05]0008/0000 (CurrentLevel) - 160U
06:56:10.947 MTR: >Read_Attr (   887) [06]0008/0000 (CurrentLevel) - 160U
06:56:10.960 MTR: >Read_Attr (   887) [07]0008/0000 (CurrentLevel) - 160U
06:56:10.973 MTR: >Read_Attr (   887) [09]0008/0000 (CurrentLevel) - 160U
06:56:11.126 CFG: Saved, Count 125, Bytes 4096
06:56:11.163 MTR: .Sub_Done  (      ) sub=2041
```

**Reply by ivanvachovsky** [reply]
Above log is for changing from GH.

**Reply by ivanvachovsky** [reply]
... and here is changing from the console PWM7. Pls note that I changed PWM8 from GH to 63% and  it doesn't show as 63% in Tasmota reports

```
07:04:27.642 CMD: PWM7 33
07:04:27.643 SRC: WebConsole from 192.168.1.205
07:04:27.645 CMD: Grp 0, Cmd 'PWM', Idx 7, Len 2, Pld 33, Data '33'
07:04:27.650 MQT: stat/VoiceBridge1/RESULT = {"PWM":{"PWM1":0,"PWM2":0,"PWM3":0,"PWM4":289,"PWM5":0,"PWM6":0,"PWM7":33,"PWM8":0,"PWM9":0}}
07:04:35.888 MQT: tele/VoiceBridge1/STATE = {"Time":"2024-10-14T07:04:35","Uptime":"0T10:10:08","UptimeSec":36608,"Heap":106,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"Berry":{"HeapUsed":23,"Objects":314},"POWER":"ON","Dimmer":63,"Color":"000000A100","HSBColor":"0,0,0","White":63,"CT":153,"Channel":[0,0,0,63,0],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","PWM":{"PWM1":0,"PWM2":0,"PWM3":0,"PWM4":289,"PWM5":0,"PWM6":0,"PWM7":33,"PWM8":0,"PWM9":0},"Wifi":{"AP":1,"SSId":"LITMATH","BSSId":"F4:E2:C6:24:9C:CE","Channel":1,"Mode":"HT20","RSSI":72,"Signal":-64,"LinkCount":2,"Downtime":"0T00:00:07"}}
```

**Reply by sfromis** [reply]
Technically, Tasmota does **not** provide the Dimmer feature when you split into separate non-light channels. While the Channel setup may look similar, it is not integrated with the Dimmer support in Tasmota. To be clear, I'm not sure how exactly the basic PWM support interacts with Matter, but I guess that this split may be involved.

**Reply by ivanvachovsky** [reply]
I don't understand what you mean. In my case I do have 8 PWM virtual devices only in Tasmota configuration. I say "virtual" because no real lights or any extra hardware is connected to the GPIOs. BTW there is no "dimmer" in Tasmota config menu. I use the only available PWM.

**Reply by s-hadinger** [reply]
I don't fully understand your configuration.

Can you send the output of `template` and `module` command?

Also can you send your `_matter_device.json` file?

**Reply by s-hadinger** [reply]
Also, if the PWM are purely virtual, why don't you use Virtual Matter devices? Why do you still use `PWM` command?

**Reply by sfromis** [reply]
To explain what I mean. When you have a "light" in Tasmota, consisting of 1-5 PWM channels (with varying color controls), you also get a dimmer feature, where you have the `Dimmer` command to vary the combined brightness of the light, as a single control. To be clear, the word "light" is not about whether you get visible light, but how it is processed by tasmota.

What you have instead is not a light, but separate PWM channels, and therefore the usual dimmer controls just do not apply. Instead there is the `Channel` command which acts a bit like a dumb dimmer without features like gamma correction. And the `PWM` command, but that's even further from being a "dimmer".

In general, I'd not expect any Tasmota feature intended for controlling a light, very much including dimming, to work for a bunch of separate PWM channels.

**Reply by ivanvachovsky** [reply]
@s-hadinger here is template

11:19:43.950 CMD: template
11:19:43.952 SRC: WebConsole from 192.168.1.205
11:19:43.953 CMD: Grp 0, Cmd 'TEMPLATE', Idx 1, Len 0, Pld -99, Data ''
11:19:43.959 MQT: stat/VoiceBridge1/RESULT = {"NAME":"ESP32-DevKit","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}

here is module

11:21:42.556 CMD: module
11:21:42.557 SRC: WebConsole from 192.168.1.205
11:21:42.559 CMD: Grp 0, Cmd 'MODULE', Idx 1, Len 0, Pld -99, Data ''
11:21:42.563 MQT: stat/VoiceBridge1/RESULT = {"Module":{"1":"ESP32-DevKit"}}

Not sure where to find the .json file here is the command

19:22:39.833 CMD: _matter_device.json
19:22:39.835 SRC: WebConsole from 192.168.1.205
19:22:39.837 CMD: Grp 0, Cmd 'MATTER_DEVICE', Idx 1, Len 5, Pld -99, Data '.json'

**Reply by ivanvachovsky** [reply]
@sfromis I don't see "light" or "dimmer" anywhere in Tasmota Configuration. Hence I used the closest one PWM to hold values from 0 to 100. This is all I want - to control values by voice ( GH ) and pass along over MQTT. I achieved this goal (kind of) as far as I'm getting MQTT reports like this every time I change my virtual PWMs by voice

05:02:17.197 MQT: stat/VoiceBridge1/COMMAND = {"MtrReceived":{"3":{"Ep":3,"Bri":76,"Dimmer":30,"Power":1}}}

However being a perfectionist it bothers me that there is a mess at Tasmota (devices do not show correct values) and at GH (all devices change when one does) as explained above.

**Reply by sfromis** [reply]
You very much see "Dimmer" in what you just posted now, from Matter. It is trying to control a Dimmer, not a PWM channel.

In other words, it is trying to control a "light", and as I just told you, configuring 1-5 PWM channels **is** a light in Tasmota, unless you take additional steps to decouple them from being a light. 

And the word "dimmer" is not something you configure, it is something you automatically get when you have a light, no matter how hard you try to circumvent it like by using PWM commands instead of how they are supposed to work.

**Reply by ivanvachovsky** [reply]
@sfromis It all makes sense now that I understand more of what you're saying. Actually Matter created only 1 Matter device with 8 PWM devices set in Tasmota. I added the rest manually selecting "Light1 Dimmer". Should I choose something else to better fit my goal?

**Reply by sfromis** [reply]
Well, my point is that a light is **one** dimmable feature (or can be split into two). You cannot just have the number you want. What you *can* do is to split the the PWM1-5 gpios to have a separate channel each, using `SetOption68`, but those are not lights and cannot be controlled with the `Dimmer` command. Therefore, you should not expect them to be seen to Matter as dimmers, unless you can do something with virtual channels (which I do not know much about).

With multiple PWM channels, PWM1-5 become a light, unless you configure to use basic PWM instead (to be controlled by the PWM command). If you do not use basic PWM mode, the 5 PWM channels of the light can be split into 5 channels (which do not work as lights, but does get web UI sliders), controllable by the Channel command.

Since you use ESP32, you have the Berry language available, where you *might* be able create some sort of bridging feature to transform between Matter and "raw" PWM channels. Not a topic I really can suggest anything specific about.

**Reply by ivanvachovsky** [reply]
@sfromis Thanks. I tried SetOption68 both 1 and 0 and in both cases the PWM devices seem to be lumped into one light.

**Reply by sfromis** [reply]
No, if you enable `SetOption68`, PWM channels will not be combined for one light.

**Comment by ivanvachovsky** [comment]
Well here is the log proving otherwise. First the values of PWM do not take what Channel command asks for and second there is "Color" in the end meaning that Tasmota somehow treats them as light IMHO.

08:41:35.274 MQT: stat/VoiceBridge1/RESULT = {"SetOption68":"ON"}
08:41:50.533 MTR: <Sub_Alive (  9639) sub=37625
08:41:56.693 CMD: Channel 12,34,45,56,67
08:41:56.695 SRC: WebConsole from 192.168.1.205
08:41:56.697 CMD: Grp 0, Cmd 'CHANNEL', Idx 1, Len 14, Pld 12, Data '12,34,45,56,67'
08:41:56.703 MQT: stat/VoiceBridge1/RESULT = {"POWER1":"ON","Channel1":12,"POWER2":"OFF","Channel2":100,"POWER3":"OFF","Channel3":98,"POWER4":"OFF","Channel4":0,"POWER5":"OFF","Channel5":0,"Color":"1FFFF80000"}

**Reply by sfromis** [reply]
That log is not proving otherwise, in any way, on the contrary. It shows 5 channels, which is when it is **not** a combined "light". That "Color" value you see is not a color, but another representation of the Channel values; not really useful.

You should not try to mix up different types of control, but configure for *either* light, channels, or PWM, and only used the commands appropriate for each type, to avoid confusion. However, don't assume direct Matter integration except for a light setup. In other words, don't expect "PWM devices" (as you call it) to work with Matter just so.

**Comment by ivanvachovsky** [comment]
Please bear with me as I'm the rookie here but isn't this one a legitimate Tasmota command (numbers are random do not mean anything)

Channel 12,34,45,56,67

If yes, than why didn't it set up the properties of the 5 "independent" PMW.

**Reply by sfromis** [reply]
No, that is not following the [Channel command](https://tasmota.github.io/docs/Commands/#channel) syntax, hence invalid. To have lots of features on a tiny MCU, Tasmota will generally not try to validate commands and report syntax errors, but stop after having parsed some of it.

Valid would be separate commands, one per independent channel:
`Backlog0 Channel1 12;Channel2 34;Channel3 45;Channel4 56;Channel5 67`
(With `Backlog` for stacking multiple commands at the same time).

**Comment by ivanvachovsky** [comment]
Thanks.
